### PR TITLE
[GEOS-9385] GWC initialization thrashes CPU and filesystem at startup 

### PR DESCRIPTION
**Pull request by groldan**
See [GEOS-9385](https://osgeo-org.atlassian.net/browse/GEOS-9385) for a more in depth description of the issue and how this PR improves it.

* Avoid `DefaultTileLayerCatalog` initialization thrashing the filesystem `DefaultTileLayerCatalog.initialize()` loads `GeoServerTileLayerInfo` objects from the gwc-layers datadirectory sub folder concurrently, on the common pool, thrashing the file system if there are many cores.    
Instead of using `ForkJoinPool.commonPool()`, set up a short-lived `ForkJoinPool` with limited parallelism, and make it respect the `org.geoserver.catalog.loadingThreads` system property, following the same settings as `AsynchResourceIterator`.
* Reduce resource usage of `GWCInitializer`
* Overcome XStream poor concurrency at `DefaultTileLayerCatalog` parallel initialization by using one `XStream` instance per thread.
* Proper handling of DefaultTileLayerCatalog life cycle

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [ ] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [ ] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [ ] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [ ] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

File: src/gwc/src/main/java/org/geoserver/gwc/layer/DefaultTileLayerCatalog.java [hunk 1]
```diff
@@ -172,22 +231,69 @@ public void initialize() {
...
-        if (!initialized) {
-            initialize();
-        }
+    /**
+     * Precondition check all public methods should make before proceeding to ensure they've been
+     * called on an initialized state
+     *
+     * @throws IllegalStateException if this layer catalog has not been initialized yet
+     */
+    private void checkInitialized() {
```
Comment (aaime) [inline src/gwc/src/main/java/org/geoserver/gwc/layer/DefaultTileLayerCatalog.java]: Hmm... this method changed from "check if initialized and if not initialize" to "check if initialized and throw an exception if not". So I'm guessing that's why there is an explicit call to initialize elsewhere. Not a big deal, just wondering.

File: src/gwc/src/main/java/org/geoserver/gwc/layer/DefaultTileLayerCatalog.java [hunk 2]
```diff
@@ -49,11 +56,53 @@
...
+
+    private static final int INITIALIZATION_PARALLELISM;
+    /**
+     * Thread local of XStream used during initialization parallel execution, to circumvent the
+     * terrible concurrency of XStream
+     */
+    private static ThreadLocal<XStream> INITIALIZATION_SERIALIZER = new ThreadLocal<>();
+
+    static {
+        String value = GeoServerExtensions.getProperty("org.geoserver.catalog.loadingThreads");
```
[the remark this answers is not in the document]
Comment (groldan) [reply src/gwc/src/main/java/org/geoserver/gwc/layer/DefaultTileLayerCatalog.java]: IMHO `AsynchResourceIterator` is too much of an implementation detail and should not be public at all. How about moving the constant to `GeoServerLoader` instead? is could arguably be a public static method too, given it's gonna be reused at reload(), the system property value may change, say, as a hint for reload? In any case I followed your advise for now, but hold on on reviewing, I've a couple test failures to fix yet.

[the remark this answers is not in the document]
Comment (aaime) [reply src/gwc/src/main/java/org/geoserver/gwc/layer/DefaultTileLayerCatalog.java]: Moving it to GeoServer loader: :+1:

File: src/gwc/src/main/java/org/geoserver/gwc/layer/DefaultTileLayerCatalog.java (hunk 1 above)
Comment (groldan) [reply src/gwc/src/main/java/org/geoserver/gwc/layer/DefaultTileLayerCatalog.java]: Correct. DefaultTileLayerCatalog initialization is and shall be managed by the lifecycle management methods (reset() and initialize()). `checkInitialized()` repurposed as public methods precondition check.

**Comment by groldan** [comment]
This should be ready to go in tandem with GeoWebCache/geowebcache/pull/797

746dc1642c2b1e1dffe141a6726e4d00b0a0a26d fixes the order dependency in a test case
f4fcdd4da5bec8d99df167b6f0ba86228414d30a fixes an issue with the UI for parameter filters, which worked by accident as a side effect of `CaseNormalizer.equals(Object)` always returning `false`

**Comment by aaime** [comment]
And merged here too.

**Comment by aaime** [comment]
Sigh, no, I have to revert this one, it breaks startup with the default data directory (hence, a regression).... I tested other data dirs, but not the release one:

````
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'gwcFacade' defined in URL [jar:file:/tmp/geoserver-2.17-SNAPSHOT/webapps/geoserver/WEB-INF/lib/gs-gwc-2.17-SNAPSHOT.jar!/applicationContext.xml]: Cannot resolve reference to bean 'DiskQuotaMonitor' while setting constructor argument; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'DiskQuotaMonitor' defined in URL [jar:file:/tmp/geoserver-2.17-SNAPSHOT/webapps/geoserver/WEB-INF/lib/gs-gwc-2.17-SNAPSHOT.jar!/geowebcache-diskquota-context.xml]: Invocation of init method failed; nested exception is java.lang.IllegalStateException: Could not locate a layer or layer group with id LayerInfoImpl--197d1929:16dcfe04a86:-7ff4 within GeoServer configuration, the GWC configuration seems to be out of synch
	at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveReference(BeanDefinitionValueResolver.java:378)
	at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveValueIfNecessary(BeanDefinitionValueResolver.java:110)
	at org.springframework.beans.factory.support.ConstructorResolver.resolveConstructorArguments(ConstructorResolver.java:674)
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:188)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1308)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1154)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:538)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:498)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:846)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:863)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:546)
	at org.springframework.web.context.ContextLoader.configureAndRefreshWebApplicationContext(ContextLoader.java:400)
	at org.springframework.web.context.ContextLoader.initWebApplicationContext(ContextLoader.java:291)
	at org.springframework.web.context.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:103)
	at org.geoserver.platform.GeoServerContextLoaderListener.contextInitialized(GeoServerContextLoaderListener.java:22)
	at org.eclipse.jetty.server.handler.ContextHandler.callContextInitialized(ContextHandler.java:957)
	at org.eclipse.jetty.servlet.ServletContextHandler.callContextInitialized(ServletContextHandler.java:553)
	at org.eclipse.jetty.server.handler.ContextHandler.startContext(ContextHandler.java:922)
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:365)
	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1497)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1459)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:852)
	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:278)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:545)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.deploy.bindings.StandardStarter.processBinding(StandardStarter.java:46)
	at org.eclipse.jetty.deploy.AppLifeCycle.runBindings(AppLifeCycle.java:192)
	at org.eclipse.jetty.deploy.DeploymentManager.requestAppGoal(DeploymentManager.java:502)
	at org.eclipse.jetty.deploy.DeploymentManager.addApp(DeploymentManager.java:151)
	at org.eclipse.jetty.deploy.providers.ScanningAppProvider.fileAdded(ScanningAppProvider.java:172)
	at org.eclipse.jetty.deploy.providers.WebAppProvider.fileAdded(WebAppProvider.java:416)
	at org.eclipse.jetty.deploy.providers.ScanningAppProvider$1.fileAdded(ScanningAppProvider.java:65)
	at org.eclipse.jetty.util.Scanner.reportAddition(Scanner.java:610)
	at org.eclipse.jetty.util.Scanner.reportDifferences(Scanner.java:529)
	at org.eclipse.jetty.util.Scanner.scan(Scanner.java:392)
	at org.eclipse.jetty.util.Scanner.doStart(Scanner.java:313)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.deploy.providers.ScanningAppProvider.doStart(ScanningAppProvider.java:145)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.deploy.DeploymentManager.startAppProvider(DeploymentManager.java:576)
	at org.eclipse.jetty.deploy.DeploymentManager.doStart(DeploymentManager.java:238)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:167)
	at org.eclipse.jetty.server.Server.start(Server.java:418)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:119)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:113)
	at org.eclipse.jetty.server.Server.doStart(Server.java:382)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.xml.XmlConfiguration$1.run(XmlConfiguration.java:1572)
	at org.eclipse.jetty.xml.XmlConfiguration$1.run(XmlConfiguration.java:1512)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1511)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.jetty.start.Main.invokeMain(Main.java:220)
	at org.eclipse.jetty.start.Main.start(Main.java:490)
	at org.eclipse.jetty.start.Main.main(Main.java:77)
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'DiskQuotaMonitor' defined in URL [jar:file:/tmp/geoserver-2.17-SNAPSHOT/webapps/geoserver/WEB-INF/lib/gs-gwc-2.17-SNAPSHOT.jar!/geowebcache-diskquota-context.xml]: Invocation of init method failed; nested exception is java.lang.IllegalStateException: Could not locate a layer or layer group with id LayerInfoImpl--197d1929:16dcfe04a86:-7ff4 within GeoServer configuration, the GWC configuration seems to be out of synch
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1745)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:576)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:498)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveReference(BeanDefinitionValueResolver.java:367)
	... 62 more
Caused by: java.lang.IllegalStateException: Could not locate a layer or layer group with id LayerInfoImpl--197d1929:16dcfe04a86:-7ff4 within GeoServer configuration, the GWC configuration seems to be out of synch
	at org.geoserver.gwc.layer.GeoServerTileLayer.lambda$getPublishedInfo$0(GeoServerTileLayer.java:385)
	at java.util.concurrent.atomic.AtomicReference.accumulateAndGet(AtomicReference.java:227)
	at org.geoserver.gwc.layer.GeoServerTileLayer.getPublishedInfo(GeoServerTileLayer.java:375)
	at org.geoserver.gwc.layer.GeoServerTileLayer.getLatLonBbox(GeoServerTileLayer.java:332)
	at org.geoserver.gwc.layer.GeoServerTileLayer.computeGridSubsets(GeoServerTileLayer.java:916)
	at org.geoserver.gwc.layer.GeoServerTileLayer.lambda$gridSubsets$1(GeoServerTileLayer.java:870)
	at java.util.concurrent.atomic.AtomicReference.accumulateAndGet(AtomicReference.java:227)
	at org.geoserver.gwc.layer.GeoServerTileLayer.gridSubsets(GeoServerTileLayer.java:866)
	at org.geoserver.gwc.layer.GeoServerTileLayer.getGridSubsets(GeoServerTileLayer.java:847)
	at org.geowebcache.diskquota.LayerCacheInfoBuilder.findOnDiskTileSets(LayerCacheInfoBuilder.java:154)
	at org.geowebcache.diskquota.LayerCacheInfoBuilder.buildCacheInfo(LayerCacheInfoBuilder.java:105)
	at org.geowebcache.diskquota.DiskQuotaMonitor.launchCacheInfoGatheringThreads(DiskQuotaMonitor.java:405)
	at org.geowebcache.diskquota.DiskQuotaMonitor.startUpInternal(DiskQuotaMonitor.java:264)
	at org.geowebcache.diskquota.DiskQuotaMonitor.startUp(DiskQuotaMonitor.java:221)
	at org.geowebcache.diskquota.DiskQuotaMonitor.afterPropertiesSet(DiskQuotaMonitor.java:178)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1804)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1741)
	... 69 more
````

Probably a misconfiguration, but the old code was somehow tolerant to it. There is plenty of broken data dirs around, an issue like this one should not prevent startup.

**Comment by aaime** [comment]
Never mind, seems it was an issue in my local data directory that can be reproduced even reverting the change in this PR... should not happen, but it's not the PR's fault it seems.

**Pull request merged**